File: leetcode/python/2352_equal_row_and_column_pairs.py

from collections import defaultdict


class Solution:
    def equalPairs(self, grid: list[list[int]]) -> int:
        rows = defaultdict(int)
        cols = defaultdict(int)
        n = len(grid)
        for i in range(n):
            rows[tuple(grid[i])] += 1
            col = []
            for j in range(n):
                col.append(grid[j][i])
            cols[tuple(col)] += 1
        res = 0
        for k, v in rows.items():
            if k in cols:
                res += v * cols[k]
        return res
        

File: leetcode/rust/src/2352_equal_row_and_column_pairs.rs
use std::collections::HashMap;

pub fn equal_pairs(grid: Vec<Vec<i32>>) -> i32 {
    let n = grid.len();
    let mut count = 0i32;

    let mut rows = HashMap::new();
    for row in &grid {
        rows.entry(row)
            .and_modify(|f| *f += 1)
            .or_insert(1);
    }

    for col in 0..n {
        let mut key = Vec::new();
        for row in 0..n {
            key.push(grid[row][col]);
        }
        count += rows.get(&key).unwrap_or(&0);
    }

    count
}
